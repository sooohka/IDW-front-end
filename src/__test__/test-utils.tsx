/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
