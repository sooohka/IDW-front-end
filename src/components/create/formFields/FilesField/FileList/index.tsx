import React, { useLayoutEffect, useRef, useState } from "react";
import { ReactComponent as Sort } from "../../../../../assets/icons/sort-up-solid.svg";
import Text from "../../../../common/Text";
import * as S from "./Style";

const FileList: React.FC = ({ children }) => {
  const [isFolded, setIsFolded] = useState(true);
  const bottomEl = useRef<HTMLDivElement>(null);

  const handleFoldBtnClick = () => {
    setIsFolded((prev) => !prev);
  };

  useLayoutEffect(() => {
    if (isFolded) bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "end" });
    else bottomEl.current?.scrollIntoView({ behavior: "smooth", block: "center" });
  }, [isFolded]);

  return (
    <S.FileList>
      <S.Header onClick={handleFoldBtnClick}>
        <S.Title>파일들</S.Title>
        <S.IconWrapper isFolded={isFolded}>
          <Sort width={30} height={30} />
        </S.IconWrapper>
      </S.Header>
      <S.Files isFolded={isFolded}>
        {React.Children.count(children) === 0 ? (
          <div style={{ textAlign: "center" }}>
            <Text bold fontSize='strongBody' text='파일이 없습니다' />
          </div>
        ) : (
          children
        )}
      </S.Files>
      <div ref={bottomEl} style={{ display: "flex", justifyContent: "flex-end" }}>
        <button type='button' onClick={handleFoldBtnClick}>
          Fold
        </button>
      </div>
    </S.FileList>
  );
};

export default FileList;
