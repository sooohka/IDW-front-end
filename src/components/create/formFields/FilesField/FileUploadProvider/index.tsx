import React, { createContext, useCallback, useContext, useReducer } from "react";

// actions
const DELETE_FILE = "file/deleteFile" as const;
const APPEND_FILES = "file/appendFiles" as const;
const UPDATE_FILE = "file/updateFile" as const;
const SET_ACCEPTING = "file/setIsAccepting" as const;

// action creators
const appendFiles = (files: TargetFile[]) => ({ type: APPEND_FILES, payload: { files } });
const updateFiles = (file: TargetFile) => ({ type: UPDATE_FILE, payload: { file } });
const deleteFile = (id: string) => ({ type: DELETE_FILE, payload: { id } });
const setIsAccepting = (isAccepting: boolean) => ({
  type: SET_ACCEPTING,
  payload: { isAccepting },
});

type Action =
  | ReturnType<typeof appendFiles>
  | ReturnType<typeof updateFiles>
  | ReturnType<typeof deleteFile>
  | ReturnType<typeof setIsAccepting>;

const initState: { files: TargetFile[]; isAccepting: boolean } = { files: [], isAccepting: false };
const reducer = (state = initState, action: Action): typeof initState => {
  switch (action.type) {
    case APPEND_FILES: {
      const { files } = action.payload;
      return { ...state, files: [...state.files, ...files] };
    }
    case UPDATE_FILE: {
      const { file } = action.payload;
      return { ...state, files: state.files.map((v) => (v.id === file.id ? file : v)) };
    }
    case DELETE_FILE: {
      const { id } = action.payload;
      return { ...state, files: state.files.filter((file) => file.id !== id) };
    }
    case SET_ACCEPTING: {
      const { isAccepting } = action.payload;
      return { ...state, isAccepting };
    }
    default:
      return state;
  }
};

interface FileUploadContextType {
  isAccepting: boolean;
  files: TargetFile[];
  handleFilesAppend: (files: TargetFile[]) => void;
  handleFileDelete: (id: string) => void;
  handleFileUpdate: (fils: TargetFile) => void;
  handleSetIsAccepting: (isAccepting: boolean) => void;
}

const FileUploadContext = createContext<FileUploadContextType>({
  isAccepting: false,
  files: [],
  handleFileDelete: () => {},
  handleFileUpdate: () => {},
  handleFilesAppend: () => {},
  handleSetIsAccepting: () => {},
});

const FileUploadProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);

  const handleFilesAppend = useCallback((files: TargetFile[]) => {
    dispatch(appendFiles(files));
  }, []);
  const handleFileDelete = useCallback((id: string) => {
    dispatch(deleteFile(id));
  }, []);
  const handleSetIsAccepting = useCallback((isAccepting: boolean) => {
    dispatch(setIsAccepting(isAccepting));
  }, []);
  const handleFileUpdate = useCallback((file: TargetFile) => {
    dispatch(updateFiles(file));
  }, []);

  const value = {
    files: state.files,
    isAccepting: state.isAccepting,
    handleFileDelete,
    handleFilesAppend,
    handleSetIsAccepting,
    handleFileUpdate,
  };
  return <FileUploadContext.Provider value={value}>{children}</FileUploadContext.Provider>;
};

const useFileUploadContext = () => useContext(FileUploadContext);

export { useFileUploadContext };
export default FileUploadProvider;
