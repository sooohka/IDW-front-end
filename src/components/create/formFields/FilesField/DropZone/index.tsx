import { FileRejection, useDropzone } from "react-dropzone";
import * as uuid from "uuid";
import UploadIcon from "../../../../../assets/icons/cloud-upload-alt-solid.svg";
import defaultTheme from "../../../../../style/theme";
import { useFileUploadContext } from "../FileUploadProvider";
import * as S from "./Style";

const DropZone: React.FC = () => {
  const { files, isAccepting, handleSetIsAccepting, handleFilesAppend } = useFileUploadContext();
  const onDragEnter = () => {
    handleSetIsAccepting(true);
  };

  const onDragLeave = () => {
    handleSetIsAccepting(false);
  };

  const onDrop = (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
    handleSetIsAccepting(true);
    const accFiles = acceptedFiles.map((acceptedFile) => ({
      file: acceptedFile,
      errors: [],
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));

    const rejFiles = rejectedFiles.map((rejectedFile) => ({
      file: rejectedFile.file,
      errors: rejectedFile.errors,
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));
    handleFilesAppend([...accFiles, ...rejFiles]);
    handleSetIsAccepting(false);
  };

  const handleValidation = (file: File) => {
    if (files.find((imageFile) => imageFile.file.name === file.name)) {
      return {
        code: "duplicate-file",
        message: "중복된 파일은 허용이 안됩니다.",
      };
    }
    if (file.size / 1024 / 1024 > 5)
      return {
        code: "file-too-large",
        message: `파일은 5mb를 넘을 수 없습니다`,
      };
    return null;
  };
  const { getRootProps, getInputProps } = useDropzone({
    accept: "image/*",
    maxFiles: 50,
    onDrop,
    onDragEnter,
    onDragLeave,
    validator: handleValidation,
  });
  return (
    <S.DropZone {...getRootProps()} isAccepting={isAccepting}>
      <input data-testid='dropzone-input' {...getInputProps()} />
      <UploadIcon fill={defaultTheme.colors.primary} width={50} height={50} />
      <p>Drag and Drop or click here to upload</p>
    </S.DropZone>
  );
};

export default DropZone;
