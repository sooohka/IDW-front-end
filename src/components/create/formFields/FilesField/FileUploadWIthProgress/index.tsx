import React, { useCallback, useEffect, useState } from "react";
import { ReactComponent as UploadFail } from "../../../../../assets/icons/file-excel-regular.svg";
import { ReactComponent as EmptyFile } from "../../../../../assets/icons/file-image-regular.svg";
import { ReactComponent as Spinner } from "../../../../../assets/icons/spinner.svg";
import { handleAwsUpload, handleCloudinaryUpload } from "../../../../../utils/lib/file";
import HelperText from "../../../../common/HelperText";
import Img from "../../../../common/Img";
import ProgressBar from "../../../../common/ProgressBar";
import XButton from "../../../../common/XButton";
import { useFileUploadContext } from "../FileUploadProvider";
import * as S from "./Style";

const handleUpload =
  process.env.NODE_ENV === "development" ? handleCloudinaryUpload : handleAwsUpload;
interface IProps {
  imageFile: TargetFile;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
}

const FileUploadWithProgress: React.FC<IProps> = ({ imageFile, setIsFileUploading }) => {
  const { handleFileDelete, handleFileUpdate } = useFileUploadContext();
  const { errors, file, id, isSubmitted, url } = imageFile;
  const [progress, setProgress] = useState(0);

  const uploadFile = useCallback(async () => {
    setIsFileUploading(true);
    try {
      let uploadedFile: TargetFile;
      if (process.env.NODE_ENV === "test") uploadedFile = { ...imageFile };
      else uploadedFile = await handleUpload(imageFile, setProgress);
      handleFileUpdate(uploadedFile);
    } catch (e) {
      // retry
    }
    setIsFileUploading(false);
  }, [handleFileUpdate, imageFile, setIsFileUploading]);

  useEffect(() => {
    if (isSubmitted) return;
    if (errors.length) return;
    uploadFile();
  }, [errors, isSubmitted, uploadFile]);

  const FileImage = () =>
    isSubmitted ? (
      <Img width='50px' height='50px' src={url} alt={file.name} />
    ) : (
      <UploadFail width={50} height={50} color='red' />
    );

  return (
    <S.FileUploadWithProgress>
      {errors.length > 0 || progress !== 100 ? (
        <EmptyFile data-testid='empty-file-svg' width={50} height={50} />
      ) : (
        <FileImage />
      )}
      <S.ProgressBar>
        <S.ProgressWrapper>
          <ProgressBar title={file.name} hasError={errors.length > 0} progress={progress} />
          {errors.length > 0 || progress === 100 ? (
            <XButton data-testid='x-button' onClick={() => handleFileDelete(id)} />
          ) : (
            <Spinner role='status' />
          )}
        </S.ProgressWrapper>
        {progress === 100 ||
          (errors.length > 0 && (
            <HelperText hasError={errors.length > 0}>
              {"submitted" && errors[0]?.message}
            </HelperText>
          ))}
      </S.ProgressBar>
    </S.FileUploadWithProgress>
  );
};

export default FileUploadWithProgress;
