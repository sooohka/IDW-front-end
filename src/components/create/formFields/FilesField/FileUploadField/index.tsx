import React, { useEffect } from "react";
import DropZone from "../DropZone";
import * as S from "../DropZone/Style";
import FileList from "../FileList";
import { useFileUploadContext } from "../FileUploadProvider";
import FileUploadWithProgress from "../FileUploadWIthProgress";

interface IProps {
  isFileUploading: boolean;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
  handleFilesChange: (
    value: {
      url: string;
      name: string;
    }[],
  ) => void;
}
const FileUploadField: React.FC<IProps> = ({
  isFileUploading,
  setIsFileUploading,
  handleFilesChange,
}) => {
  const { files, handleFileDelete } = useFileUploadContext();

  useEffect(() => {
    if (isFileUploading) return;
    if (files.some((file) => !file.isSubmitted || file.errors.length > 0)) return;
    const formedFiles = files.map((imageFile) => ({
      name: imageFile.file.name,
      url: imageFile.url,
    }));
    handleFilesChange(formedFiles);
  }, [files, handleFilesChange, isFileUploading]);
  return (
    <S.FileUploadField>
      <DropZone />
      <FileList>
        {files.map((file) => (
          <FileUploadWithProgress
            setIsFileUploading={setIsFileUploading}
            imageFile={file}
            key={file.id}
          />
        ))}
      </FileList>
    </S.FileUploadField>
  );
};
export default FileUploadField;
