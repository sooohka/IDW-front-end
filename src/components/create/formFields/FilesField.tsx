import { FormikErrors } from "formik";
import React, { memo, useCallback } from "react";
import styled from "styled-components";
import HelperText from "../../common/HelperText";
import FileUploadField from "../FileUploadField";

const StyledField = styled.div`
  display: flex;
  flex-direction: column;
`;

const FieldTitle = styled.span`
  display: flex;
  flex-direction: column;
  font-weight: bold;
  letter-spacing: 1px;
  font-size: ${({ theme }) => theme.fonts.label};
`;

interface IProps {
  name: keyof CreateFormValues;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
  error?:
    | string
    | string[]
    | FormikErrors<{
        url: string;
        name: string;
      }>[];
  handleFilesChange: (name: string) => (value: CreateFormValues["files"]) => void;
}

const FilesField: React.FC<IProps> = ({ handleFilesChange, error, name, setIsFileUploading }) => {
  return (
    <>
      <StyledField>
        <FieldTitle>파일</FieldTitle>
        <FileUploadField
          handleFilesChange={useCallback(() => handleFilesChange(name), [handleFilesChange, name])}
          setIsFileUploading={setIsFileUploading}
        />
      </StyledField>
      <HelperText hasError={Boolean(error)} text={error as string} />
    </>
  );
};

export default FilesField;
