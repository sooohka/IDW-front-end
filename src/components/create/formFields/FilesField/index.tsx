import { FormikErrors } from "formik";
import React, { useCallback } from "react";
import HelperText from "../../../common/HelperText";
import FileUploadField from "./FileUploadField";
import FileUploadProvider from "./FileUploadProvider";
import * as S from "./Style";

interface IProps {
  name: keyof CreateFormValues;
  isFileUploading: boolean;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
  error?:
    | string
    | string[]
    | FormikErrors<{
        url: string;
        name: string;
      }>[];
  handleFilesChange: (name: string) => (value: CreateFormValues["files"]) => void;
}

const FilesField: React.FC<IProps> = ({
  handleFilesChange: filesChange,
  error,
  name,
  isFileUploading,
  setIsFileUploading,
}) => {
  const handleFilesChange = useCallback(
    (value: CreateFormValues["files"]) => {
      return filesChange(name)(value);
    },
    [filesChange, name],
  );
  return (
    <>
      <S.Field>
        <S.FieldTitle>파일</S.FieldTitle>
        <FileUploadProvider>
          <FileUploadField
            isFileUploading={isFileUploading}
            handleFilesChange={handleFilesChange}
            setIsFileUploading={setIsFileUploading}
          />
        </FileUploadProvider>
      </S.Field>
      <HelperText hasError={Boolean(error)}>{error as string}</HelperText>
    </>
  );
};

export default FilesField;
