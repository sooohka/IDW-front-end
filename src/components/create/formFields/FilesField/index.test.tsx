import { ReactElement } from "react";
import FilesField from ".";
import { createWithThemeProvider } from "../../../../__test__/test-utils";

describe("components/create/formFields/FilesField", () => {
  let FileFieldEL: ReactElement;
  const filesFieldProps: any = {
    handleFilesChange: null,
    isFileUploading: false,
    name: "",
    setIsFileUploading: null,
    error: null,
  };

  beforeEach(() => {
    filesFieldProps.handleFilesChange = jest.fn();
    filesFieldProps.isFileUploading = jest.fn();
    filesFieldProps.name = "files";
    filesFieldProps.setIsFileUploading = jest.fn();
    filesFieldProps.error = undefined;
    FileFieldEL = <FilesField {...filesFieldProps} />;
  });
  it("renders", () => {
    const component = createWithThemeProvider(FileFieldEL);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
