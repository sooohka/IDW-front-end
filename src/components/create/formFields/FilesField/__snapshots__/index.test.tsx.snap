// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/create/formFields/FilesField renders 1`] = `
Array [
  <div
    className="sc-bkkeKt gzyHCl"
  >
    <span
      className="sc-ieecCq hxrxLM"
    >
      파일
    </span>
    <div
      className="sc-gsDKAQ hypcjQ"
    >
      <div
        className="sc-dkPtRN cUA-dwf"
        onBlur={[Function]}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <input
          accept="image/*"
          autoComplete="off"
          data-testid="dropzone-input"
          multiple={true}
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          tabIndex={-1}
          type="file"
        />
        <svg
          fill="#2d3436"
          height={50}
          width={50}
        >
          cloud-upload-alt-solid.svg
        </svg>
        <p>
          Drag and Drop or click here to upload
        </p>
      </div>
      <div
        className="sc-eCImPb idENhH"
      >
        <div
          className="sc-gKclnd jYkcZO"
          data-testid="file-list-header"
          onClick={[Function]}
        >
          <h3
            className="sc-jRQBWg cOogGM"
          >
            파일들
          </h3>
          <div
            className="sc-furwcr efWBEw"
          >
            <svg
              height={30}
              width={30}
            >
              sort-up-solid.svg
            </svg>
          </div>
        </div>
        <div
          className="sc-iCfMLu fPpHSm"
          data-testid="file-list-content"
        >
          <div
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <p
              className="sc-hKwDye cenYoF"
              color="primary"
              fontSize="strongBody"
            >
              파일이 없습니다
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <p
    className="sc-bdvvtL iDNMqj"
    role="note"
  />,
]
`;
